catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/history/Historian.java

```diff
@@ -0,0 +1,513 @@
+                                .setAttribute(new AttributeImpl(Metacard.RESOURCE_URI,
+                                        ci.getUri()));
+                        ci.getMetacard()
+                                .setAttribute(new AttributeImpl(Metacard.RESOURCE_SIZE,
+                                        ci.getSize()));
+                    } catch (IOException e) {
+                        LOGGER.warn("Could not get size", e);
+                    }
+
+                });
+    }
+
+    private List<ContentItem> versionContentItems(List<ContentItem> contentItems,
+            Map<String, Path> tmpContentPaths, MetacardVersion.Action action) {
+        return contentItems.stream()
+                .map(ContentItem::getMetacard)
+                .distinct()
+                .flatMap(mc -> getVersionedContent(mc,
+                        contentItems,
+                        action,
+                        tmpContentPaths).stream())
+                .collect(Collectors.toList());
+    }
+
+    private void setResourceUriIfMissing(List<ContentItem> newContentItems) {
+        newContentItems.stream()
+                .filter(ci -> ci.getMetacard()
+                        .getResourceURI() == null)
+                .filter(ci -> StringUtils.isBlank(ci.getQualifier()))
+                .forEach(ci -> {
+                    ci.getMetacard()
+                            .setAttribute(new AttributeImpl(Metacard.RESOURCE_URI, ci.getUri()));
+                    try {
+                        ci.getMetacard()
+                                .setAttribute(new AttributeImpl(Metacard.RESOURCE_SIZE,
+                                        ci.getSize()));
+                    } catch (IOException e) {
+                        LOGGER.warn("Could not get size of content item", e);
+                    }
+                });
+    }
+
+    private CreateResponse getVersionedMetacards(List<Metacard> metacards,
+            final MetacardVersion.Action action)
+            throws SourceUnavailableException, IngestException {
+        final List<Metacard> versionedMetacards = metacards.stream()
+                .filter(MetacardVersion::isNotVersion)
+                .map(metacard -> new MetacardVersion(metacard, action, SecurityUtils.getSubject()))
+                .collect(Collectors.toList());
+
+        if (versionedMetacards.isEmpty()) {
+            return null;
+        }
+
+        return getSystemSubject().execute(() -> catalogProvider().create(new CreateRequestImpl(
+                versionedMetacards)));
+    }
+
+    private List<ContentItem> getVersionedContent(Metacard root, List<ContentItem> contentItems,
+            MetacardVersion.Action versionAction, Map<String, Path> tmpContentPaths) {
+        String id = root.getId();
+        MetacardVersion rootVersion = new MetacardVersion(root,
+                versionAction,
+                SecurityUtils.getSubject());
+
+        List<ContentItem> relatedContent = contentItems.stream()
```

Comment:
You only use `relatedContent` as a filtered list for your processing of `rootItem` and `derivedContent`. Instead of the expense of creating the collection, you could just do...
```java
Supplier<Stream<ContentItem> relatedContent = () -> contentItems.stream()
    .filter(ci -> ci.getId().equals(id));

ContentItem rootItem = relatedContent.get().filter()...
```